File: C/2_recursion-and-recurrence/the-longest-path/memoization.c
#include <stdio.h>
#include <string.h>

int getTheLongest(int[][100], int, int, int, int[][100]);
int max(int, int);

int main(void)
{
    int n;
    scanf("%d", &n);

    int t[100][100], rec[100][100];
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < i + 1; j++)
        {
            scanf("%d", &t[i][j]);
        }
    }
    memset(rec, -1, sizeof(rec));
    printf("%d\n", getTheLongest(t, n, 0, 0, rec));
    return 0;
}

int getTheLongest(int t[][100], int n, int i, int j, int rec[][100])
{
    if (i == n - 1)
    {
        return t[i][j];
    }
    if (rec[i][j] == -1)
    {
        rec[i][j] = max(getTheLongest(t, n, i + 1, j, rec), getTheLongest(t, n, i + 1, j + 1, rec)) + t[i][j];
    }
    return rec[i][j];
}

int max(int x, int y)
{
    return x < y ? y : x;
}

File: JavaScript/2_recursion-and-recurrence/the-longest-path/memoization.js
'use strict'

const readline = require('readline')
const rl = readline.createInterface({ input: process.stdin })

let n = 0
let t = new Array(),
    rec = new Array()

rl.on('line', input => {
    if (n === 0) {
        n = parseInt(input)
    } else {
        const values = input.split(' ')
        const len = values.length
        for (let i = 0; i < len; i++) {
            values[i] = parseInt(values[i])
        }

        t[len - 1] = values
        rec[len - 1] = new Array(len).fill(-1)

        if (t.length === n) {
            console.log(getTheLongest(0, 0))
            rl.close()
        }
    }
})

function getTheLongest(i, j) {
    if (i === n - 1) {
        return t[i][j]
    }
    if (rec[i][j] === -1) {
        rec[i][j] = max(getTheLongest(i + 1, j, rec), getTheLongest(i + 1, j + 1, rec)) + t[i][j]
    }
    return rec[i][j]
}

const max = (x, y) => x < y ? y : x
